music_assistant/providers/airplay/airplay2.py

```diff
@@ -0,0 +1,511 @@
+        if self._cli_proc and not self._cli_proc.closed:
+            await self._cli_proc.close(True)
+        if self._ffmpeg_proc and not self._ffmpeg_proc.closed:
+            await self._ffmpeg_proc.close(True)
+        self.player.set_state_from_stream(state=PlaybackState.IDLE, elapsed_time=0)
+
+    async def write_chunk(self, chunk: bytes) -> None:
+        """Write a (pcm) audio chunk."""
+        if self._stopped:
+            raise RuntimeError("Stream is already stopped")
+        await self._started.wait()
+        assert self._ffmpeg_proc
+        await self._ffmpeg_proc.write(chunk)
+
+    async def write_eof(self) -> None:
+        """Write EOF."""
+        if self._stopped:
+            raise RuntimeError("Stream is already stopped")
+        await self._started.wait()
+        assert self._ffmpeg_proc
+        await self._ffmpeg_proc.write_eof()
+
+    async def send_cli_command(self, command: str) -> None:
+        """Send an interactive command to the running CLIap2 binary."""
+        if self._stopped or not self._cli_proc or self._cli_proc.closed:
+            return
+        await self._started.wait()
+
+        if not command.endswith("\n"):
+            command += "\n"
+
+        def send_data() -> None:
+            with suppress(BrokenPipeError), open(named_pipe, "w") as f:
+                f.write(command)
+
+        named_pipe = f"/tmp/ap2-{self.active_remote_id}"  # noqa: S108
+        self.player.logger.log(VERBOSE_LOG_LEVEL, "sending command %s", command)
+        self.player.last_command_sent = time.time()
+        await asyncio.to_thread(send_data)
+
+    def start_ffmpeg_stream(self) -> None:
+        """Start (or replace) the player-specific ffmpeg stream to feed cliap2."""
+        # cancel existing ffmpeg reader task
+        if self._ffmpeg_reader_task and not self._ffmpeg_reader_task.done():
+            self._ffmpeg_reader_task.cancel()
+        if self._ffmpeg_proc and not self._ffmpeg_proc.closed:
+            self.mass.create_task(self._ffmpeg_proc.close(True))
+        # start new ffmpeg reader task
+        self._ffmpeg_reader_task = self.mass.create_task(self._ffmpeg_reader())
+
+    async def _ffmpeg_reader(self) -> None:
+        """Read audio from the audio source and pipe it to the CLIap2 process."""
+        self._ffmpeg_proc = FFMpeg(
+            audio_input="-",
+            input_format=self.session.input_format,
+            output_format=AIRPLAY_PCM_FORMAT,
+            filter_params=get_player_filter_params(
+                self.mass,
+                self.player.player_id,
+                self.session.input_format,
+                AIRPLAY_PCM_FORMAT,
+            ),
+        )
+        self._stream_bytes_sent = 0
+        await self._ffmpeg_proc.start()
+        chunksize = get_chunksize(AIRPLAY_PCM_FORMAT)
+        # wait for cliap2 to be ready
+        await asyncio.wait_for(self._started.wait(), 20)
+        async for chunk in self._ffmpeg_proc.iter_chunked(chunksize):
```

Comment:
Is this an efficient method? I suspect it is reopening the named pipe for each chunk written, and given this is a fairly IO intensive process, it might not be so efficient. Perhaps we should open the named pipe just before starting cliap2 and then have a member function to write to the named pipe using the file descriptor for the open pipe, and then only try to reopen the pipe upon a `BrokenPipeError`?
I tried using the elegant method of using extra args to the `ffmpeg` process as suggested by @marcelveldt, but this did not work. The reason was that the `chunk` was not sent to `mass` (for it to then forward to cliap2) and as a result, `mass` couldn't keep track of the `_stream_bytes_sent` and `_total_bytes_sent` values.